### DISCUSSION TITLE
Issue with file upload batch requests

### DISCUSSION THREAD

**Discussion by TheCynicalBadger** [microsoftgraph/msgraph-sdk-dotnet#2538] [Q&A]
There is a closed issue related to this question, #2089. I don't know if this is still a graph issue or not though, the problem seems to arise with the URI encoding before the batch request is made. I can get requests to work on graph explorer but only if the '!' character of the url in graph is not encoded to %21.
If you submit the drive ID with the '!' character the endpoint works and returns a 201 Created. If you encode the '!' with %21 like the request builder does it returns a 501 Not Implemented.
Any thoughts? Not sure if this is an SDK or a graph issue.
![image](https://github.com/microsoftgraph/msgraph-sdk-dotnet/assets/32382766/8147658a-0c74-4708-92f7-7157827e84db)
![image](https://github.com/microsoftgraph/msgraph-sdk-dotnet/assets/32382766/ceeff7dd-e4fe-4592-acee-ea9f4ecca8f1)

**Comment by andrueastman** [comment]
Thanks for raising this @TheCynicalBadger 

It seems that the API may be doing the work of encoding the URL from the json body so it would be unnecessary/incorrect to have the URL encoded from the client side while its in the json body(technically the API makes the HTTP requests for us).

Any chance you can confirm if you can do this successfully while using the SDK?

**Reply by TheCynicalBadger** [reply]
@andrueastman Bumping this to see if you had a chance to look into it. I tried to intercept the request before it got sent via the SDK but was unsuccessful. For now I've resorted to using JSON requests or looping through a list of IFormFile and performing a single POST for each. Thanks

**Reply by andrueastman** [reply]
@TheCynicalBadger I believe the issue lies at the code at the link below that uses the URI class to get the path and from the message. 

https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/blob/07584bb61cc4298d75fad71c9b5e5c7d14c3b01c/src/Microsoft.Graph.Core/Requests/Content/BatchRequestContent.cs#L328

Any chance you'd be willing to submit a PR to fix this? I believe you can add a test here to replicate this scenario to ensure the path section is not encoded...

https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/blob/07584bb61cc4298d75fad71c9b5e5c7d14c3b01c/tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Content/BatchRequestContentTests.cs#L594

**Reply by TheCynicalBadger** [reply]
@andrueastman I'd be willing to make a PR - If I have some free time later this week I'll see if I can get something put together. Thanks for the source links

**Comment by TheCynicalBadger** [comment]
@andrueastman Thanks for the reply.

I probably should have mentioned this in my original post. The issue comes from the GraphClient().Drives[driveId].Items[itemId].ItemWithPath(fileName).Content.ToPutRequestInformation(content)
When I used the SDK I get the 501 error and I haven't had time to figure out how to intercept the URI before it gets posted with the batch request. When I call the ToPutRequestInformation(content) the getter for the URI encodes it which returns that 501 response.

I have been unable to get the SDK to batch post files.
